File: leetcode_java/657.RobotReturnToOrigin.java
package leetcode_java;

class Solution {
    // ## Approach-1: simulate the robot's position after each move
    // ## Time: O(n)
    // ## Space: O(n), charArray
    public boolean judgeCircle(String moves) {
        int x = 0, y = 0;
        for (char c : moves.toCharArray()) {
            if (c == 'U') y += 1;
            else if (c == 'D') y -= 1;
            else if (c == 'L') x -= 1;
            else if (c == 'R') x += 1;
        }
        return x == 0 && y == 0;
    }
}

File: 657.Robot_Return_to_Origin.py
class Solution:
    def judgeCircle(self, moves: str) -> bool:
        ## Approach-1: simulate the robot's position after each move
        ## Time: O(n)
        ## Space: O(1)
        '''
        x, y = 0, 0

        for m in moves:
            if m == "U": y += 1
            elif m == "D": y -= 1
            elif m == "L": x -= 1
            elif m == "R": x += 1

        return (x, y) == (0, 0)
        '''

        ## Approach-2: 
        ## Check if count(U) == count(D) 
        ## and count(L) == count(R)
        ## Time: O(n) * 4
        ## Space: O(1)
        return moves.count("U") == moves.count("D") and moves.count("L") == moves.count("R")
    